### PR TITLE
feat: switch to target latency for live sync and add dynamic target latency feature

### LINKED ISSUE #6131: Using playback rate to help reduce rebuffering and maintain buffer health.
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes, there aren't existing issues.

**Is your feature request related to a problem? Please describe.**

We would like to further reduce the likelihood of rebuffering and increase buffer health by tweaking the playback rate of the video.

**Describe the solution you'd like**

This involves 3 changes that are all related.
- For VOD
  - `streaming.vodDynamicPlaybackRate`: when enabled, begin playback at a low playback rate (0.95x, configurable) and when the buffer is high, increase rate to 1x, if the buffer is low, decrease back to 0.95x.
- For live
  - `streaming.liveSyncPanicMode`: if enabled, When the player rebuffers, enter "panic" mode, which sets the player to the `liveSyncMinPlaybackRate` for a specified duration, say 60s by default.
  - `streaming.liveSyncDynamicTargetLatency`: if enabled, when the playback is progressing smoothly, move the target latency closer to the min latency, and if playback is having issues move it back towards the max latency.

**Describe alternatives you've considered**

It's possible to introduce other ABR algorithms, but this gives us another tool to maintain buffer health for both live and vod content and help reduce rebuffering for users.

**Additional context**

The first two features, dynamic playback rate for vod and panic mode for live sync, can both be implemented as extensions on top of what's currently there. Though, if there are pointers on how to approach this that would fit the codebase better, that would be great!
For example, for the panic mode, we'd need to store when the last rebuffering event happened so that we can check against it in the [onTimeUpate_](https://github.com/shaka-project/shaka-player/blob/291b4971f92634970e26bd6b50a331e3abe2bfad/lib/player.js#L5641-L5646) method in player.js, where should that var be stored? And for the vod dynamic playback rate, should the play rate controller own this functionality?

The third feature, dynamic target latency, is a bit trickier. Currently, live sync is set up by looking at the min and max latencies (either configured or received via DASH ServiceDescriptions) and then playback rate is set up based on that. However, for this feature, we'd need to have a third value, target latency -- which the DASH LL Spec does define, actually -- but switching from a min/max latency to a target latency would potentially be a breaking change or would potentially duplicate a lot of code (since we'd want to fork the [onTimeUpdate_](https://github.com/shaka-project/shaka-player/blob/291b4971f92634970e26bd6b50a331e3abe2bfad/lib/player.js#L5641-L5646) method depending on whether dynamic target latency is enabled or not).
While, the DASH spec, technically, says that the target latency is the only required one, I'm not sure if it's necessarily worth migrating to, given the risk just outlined. However, if y'all have thoughts on how to do that or any other alternatives, it would be nice as this feature can help reduce rebufferings.

### PR DESCRIPTION
**Pull request by gkatsev**
This switches the live sync to use target latency (which is the only required part of DASH's service descriptors) and a configurable tolerance value, which defaults to 0.5s due to a recommendation from the LL-DASH spec.
Additionally, it adds a dynamic target latency feature which shifts the current target latency towards the minimum latency value when playback is stable (based on a configurable threshold, which defaults to 60s) and also shifts the latency towards the maximum latency on rebuffering events (by the configurable increment value, which defaults to 0.5s).

Closes #6131.

### CODE REVIEW

**Comment by gkatsev** [comment]
Completely forgot about tests, I'll check on them tomorrow.

**Comment by gkatsev** [comment]
I believe this is ready for review.

**Comment by avelad** [comment]
@gkatsev Please rebase and resolve the conflicts, Thanks!

**Comment by gkatsev** [comment]
@avelad rebased and fixed conflicts!

File: test/dash/dash_parser_manifest_unit.js
Comment (avelad) [inline test/dash/dash_parser_manifest_unit.js:2586]: Thanks!

File: lib/player.js
Comment (avelad) [inline lib/player.js:682]: We should reset this value in the unload method to avoid inconsistencies between loads

**Review by avelad — CHANGES_REQUESTED** [review]
LGTM!, just one small thing needs to be fixed.

**Review by avelad — APPROVED** [review]

**Comment by joeyparrish** [comment]
OK, I'm assuming that if @avelad approved but didn't merge, he was either waiting for test results or my review.  I'll take a look now.

File: externs/shaka/player.js
Comment (gkatsev) [reply externs/shaka/player.js:1362]: I think you have a good overview of the changes. I've also previousl written a description of this here https://github.com/shaka-project/shaka-player/issues/6131#issuecomment-1906934434.

For increment to go towards min latency is half of the distance between target latency and min latency.
Additonally, it isn't just backing off in 0.5s increments, but rather in a multiple of 0.5s per each rebuffering event. So, if there's a bunch of rebuffering, we'll increment further back. https://github.com/shaka-project/shaka-player/pull/6193/files#diff-7abf4aa2639707ce10a19703663841fae6c5638809faf4dfb2f1894791f154ddR5694.

I think that the pendulum that you mention is definitely a possibility, though, I assume that folks that may be using this option will probably want to tweak their target and minimum settings away from the defaults. It probably doesn't make much difference with the default min, max, and target latency values.
Also, I'm not sure that a pendulum here is necessarily an issue. During playback we do want to react to occasions when there are rebuffering, so, back off on being close to the live edge and if playback is stable, we can move closer to the edge. Given the unpredicatiblity of the network, I think it's reasonable, particularly for an opt-in feature. Some internal tests, a feature like this decreased the live latency and reduced the rebuffering count.

Comment (gkatsev) [reply externs/shaka/player.js:1362]: maybe if the rebuffer count is high we should stick to the max latency and stop trying to move it towards min?

Comment (joeyparrish) [reply externs/shaka/player.js:1362]: >  I'm not sure that a pendulum here is necessarily an issue. During playback we do want to react to occasions when there are rebuffering, so, back off on being close to the live edge and if playback is stable, we can move closer to the edge. Given the unpredicatiblity of the network, I think it's reasonable, particularly for an opt-in feature. Some internal tests, a feature like this decreased the live latency and reduced the rebuffering count.

I hear you.  But I wonder what the behavior of this will be further from the median experience.  We may not know that until it is deployed and instrumented across a large number of sessions.  Is your "internal testing" based on real user populations in an A/B experiment?  Or just "seems good" from your fast connection at work or home?  I mean no offense in this.  I'm always forced to rely on the latter myself, but I just think we should be clear about the limitations of it when we discuss our risk/reward assessments.

> I assume that folks that may be using this option will probably want to tweak their target and minimum settings away from the defaults. It probably doesn't make much difference with the default min, max, and target latency values.

While I'm sure plenty of developers are power users who understand the nuances and can tweak all these things, I bet there are still many who don't have a deep enough knowledge or experience to do so.  Or who would Deploy an updated player without noticing that this feature is in it.  Are there better defaults we could pick?  I feel like we should assume that even if it's opt-in right now, it could become opt-out some day, and we should strive for defaults that are a good starting point for most scenarios, with low risk if possible.

> maybe if the rebuffer count is high we should stick to the max latency and stop trying to move it towards min?

Maybe so.  Maybe something like this, though I'm not certain what.  I feel strongly that something should exist to stop a rebuffering cycle if it occurs.

I feel the same about latency as resolution: better to be a little father from the bleeding edge if you can avoid buffering in most cases.

But reasonable people could certainly disagree with me on that.  I also don't have any user session data to back up my philosophies on any of this.  So I would defer to real data if anyone has some.

Comment (gkatsev) [reply externs/shaka/player.js:1362]: I totally see your concerns as well. Definitely a tricky situation trying to cater to power users while not letting others accidentally ruin their user experience.

I believe the tests were done with a network emulator and a bunch of network traces, but getting confirmation. Either way, it's definitely not the same as a full A/B experiment, though, definitely better than someone just hitting play on their laptop a bunch of times.

Would `liveSyncDynamicTargetLatencyRebufferThreshold`, where once the rebuffering count cross this boundary, it will choose the max latency and stick there without trying to push closer to the edge again assuage your concerns?

------------

Also, given that this change is essentially two in one: switch to target latency, and dynamic target latency. With shaka v5 looming, if we don't have agreement on the dynamic part, would just the target latency be able to land? Since dynamic target latency could later be added as a feature, if it doesn't happen before v5 is done. Happy to separate this PR into two if that makes it more likely for the breaking change to land in time for v5.

Comment (gkatsev) [reply externs/shaka/player.js:1362]: Confirmed my interpretation was correct regarding the test runner.

Comment (avelad) [reply externs/shaka/player.js:1362]: What would be the next action?

Comment (gkatsev) [reply externs/shaka/player.js:1362]: Oops, somehow missed your message.

For next actions, there are several options. The one that would be the least amount of work is likely to add the `liveSyncDynamicTargetLatencyRebufferThreshold` or similar option. Alternatively, I can split this PR between the refactor to target latency and the new feature of target latency. This was more necessary when v5 was scheduled soon.

I'd like us to have consensus on the next course of action before I update the PR. Is `liveSyncDynamicTargetLatencyRebufferThreshold` reasonable? Perhaps with a new name. Or maybe there are other ideas for resolving @joeyparrish's concern about the pendulum. Currently, I don't have any other ideas, but I'll spend some more time thinking about it next week. I would appreciate any other suggestions.

Comment (gkatsev) [reply externs/shaka/player.js:1362]: Another potential change is to move the dynamic target latency options into a sub config object.

Comment (avelad) [reply externs/shaka/player.js:1362]: Yes, I think it's better to split it into two PRs, it would be easier to review. As for the new setup, I'll let Joey decide.

**Comment by gkatsev** [comment]
@avelad in #6683, you mentioned that you'll have a follow-up PR related to target latency. I think it probably makes sense for this change to wait till then? If yes, do you have thoughts on timing, if not, I'm happy to prioritize getting this PR fixed up asap. Thanks!

**Comment by avelad** [comment]
> @avelad in #6683, you mentioned that you'll have a follow-up PR related to target latency. I think it probably makes sense for this change to wait till then? If yes, do you have thoughts on timing, if not, I'm happy to prioritize getting this PR fixed up asap. Thanks!

It will take 1 or 2 weeks. If you want to do that PR first I would be happy to review it.

**Comment by gkatsev** [comment]
That change involves removing the max and min latency configs? Anything else to consider? Particularly, since that would be a breaking change? One way to maintain backwards compat is to use the max/min configs to create calculate the target latency and the latency tolerance, if not provided. Then, the actual code would only consider target latency and the target latency tolerance.

**Comment by avelad** [comment]
> That change involves removing the max and min latency configs? Anything else to consider? Particularly, since that would be a breaking change? One way to maintain backwards compat is to use the max/min configs to create calculate the target latency and the latency tolerance, if not provided. Then, the actual code would only consider target latency and the target latency tolerance.

The idea is to deprecate max and min and create a target config. 
For the deprecated one see https://github.com/shaka-project/shaka-player/blob/main/lib/player.js#L3534
And I think this change should go in a separate PR so we can merge it sooner.

**Comment by gkatsev** [comment]
Ah, forgot deprecation was a thing. Yeah, definitely a separate PR.

**Comment by avelad** [comment]
@gkatsev I think this can be resumed, what do you think?

**Comment by gkatsev** [comment]
yes, i'll fix it up, probably in a separate PR

**Pull request closed without merging**
